plugins/out_azure_kusto/azure_kusto.c

```diff
@@ -0,0 +1,509 @@
+
+    if (u_conn) {
+        flb_upstream_conn_release(u_conn);
+    }
+
+    if (token) {
+        flb_sds_destroy(token);
+    }
+
+    if (body) {
+        flb_sds_destroy(body);
+    }
+
+    return NULL;
+}
+
+static int cb_azure_kusto_init(struct flb_output_instance *ins, struct flb_config *config,
+                               void *data)
+{
+    char *token;
+    int io_flags = FLB_IO_TLS;
+    int ret;
+    struct flb_azure_kusto *ctx;
+
+    /* Create config context */
+    ctx = flb_azure_kusto_conf_create(ins, config);
+    if (!ctx) {
+        flb_plg_error(ins, "configuration failed");
+        return -1;
+    }
+
+    flb_output_set_context(ins, ctx);
+
+    /* Network mode IPv6 */
+    if (ins->host.ipv6 == FLB_TRUE) {
+        io_flags |= FLB_IO_IPV6;
+    }
+
+    /* Create mutex for acquiring oauth tokens  and getting ingestion resources (they are
+     * shared across flush coroutines)
+     */
+    pthread_mutex_init(&ctx->token_mutex, NULL);
+    pthread_mutex_init(&ctx->resources_mutex, NULL);
+
+    /*
+     * Create upstream context for Kusto Ingestion endpoint
+     */
+    ctx->u = flb_upstream_create_url(config, ctx->ingestion_endpoint, io_flags, ins->tls);
+    if (!ctx->u) {
+        flb_plg_error(ctx->ins, "upstream creation failed");
+        return -1;
+    }
+
+    /* Create oauth2 context */
+    ctx->o =
+        flb_oauth2_create(ctx->config, ctx->oauth_url, FLB_AZURE_KUSTO_TOKEN_REFRESH);
+    if (!ctx->o) {
+        flb_plg_error(ctx->ins, "cannot create oauth2 context");
+        return -1;
+    }
+    flb_output_upstream_set(ctx->u, ins);
+
+    /* Get or renew the OAuth2 token */
+    token = get_azure_kusto_token(ctx);
+
+    if (!token) {
+        flb_plg_warn(ctx->ins, "token retrieval failed");
+    }
+    else {
+        flb_sds_destroy(token);
+
+        /* load or refresh ingestion resources */
+        ret = azure_kusto_load_ingestion_resources(ctx, config);
+        if (ret != 0) {
+            flb_plg_warn(ctx->ins, "ingestion resources loading failed");
+        }
+    }
+
+    return 0;
+}
+
+static int azure_kusto_format(struct flb_azure_kusto *ctx, const char *tag, int tag_len,
+                              const void *data, size_t bytes, void **out_data,
+                              size_t *out_size)
+{
+    int records = 0;
+    size_t off = 0;
+    msgpack_unpacked result;
+    msgpack_sbuffer mp_sbuf;
+    msgpack_packer mp_pck;
+    /* for sub msgpack objs */
+    int map_size;
+    struct flb_time tm;
+    struct tm tms;
+    msgpack_object root;
+    msgpack_object *obj;
+    char time_formatted[32];
+    size_t s;
+    int len;
+
+    /* output buffer */
+    flb_sds_t out_buf;
+
+    /* Create temporary msgpack buffer */
+    msgpack_sbuffer_init(&mp_sbuf);
+    msgpack_packer_init(&mp_pck, &mp_sbuf, msgpack_sbuffer_write);
+
+    /* Create array for all records */
+    records = flb_mp_count(data, bytes);
+    if (records <= 0) {
+        msgpack_sbuffer_destroy(&mp_sbuf);
```

Comment:
`msgpack_sbuffer_destroy` calls `free` passing the `data` member of the `msgpack_sbuffer` structure which was zeroed out by `msgpack_sbuffer_init` but not yet set to anything because no direct or indirect calls to `msgpack_sbuffer_write` have been made. 

Even though this should be handled by most runtimes at the moment it wouldn't hurt to move the sbuffer nad packer init calls so they are performed after this check thus ensuring we don't leave anything to luck.